yt_dlp/networking/utils.py

```diff
@@ -0,0 +1,192 @@
+from __future__ import annotations
+
+import contextlib
+import random
+import ssl
+import sys
+import urllib.parse
+import urllib.request
+
+import typing
+
+from ..compat import compat_urllib_parse_unquote_plus, compat_urlparse
+from ..dependencies import certifi
+from ..socks import ProxyType
+from ..utils import CaseInsensitiveDict, std_headers, update_url_query
+
+if typing.TYPE_CHECKING:
+    from .common import Request
+    from http.cookiejar import CookieJar
+
+
+def random_user_agent():
+    _USER_AGENT_TPL = 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/%s Safari/537.36'
+    _CHROME_VERSIONS = (
+        '90.0.4430.212',
+        '90.0.4430.24',
+        '90.0.4430.70',
+        '90.0.4430.72',
+        '90.0.4430.85',
+        '90.0.4430.93',
+        '91.0.4472.101',
+        '91.0.4472.106',
+        '91.0.4472.114',
+        '91.0.4472.124',
+        '91.0.4472.164',
+        '91.0.4472.19',
+        '91.0.4472.77',
+        '92.0.4515.107',
+        '92.0.4515.115',
+        '92.0.4515.131',
+        '92.0.4515.159',
+        '92.0.4515.43',
+        '93.0.4556.0',
+        '93.0.4577.15',
+        '93.0.4577.63',
+        '93.0.4577.82',
+        '94.0.4606.41',
+        '94.0.4606.54',
+        '94.0.4606.61',
+        '94.0.4606.71',
+        '94.0.4606.81',
+        '94.0.4606.85',
+        '95.0.4638.17',
+        '95.0.4638.50',
+        '95.0.4638.54',
+        '95.0.4638.69',
+        '95.0.4638.74',
+        '96.0.4664.18',
+        '96.0.4664.45',
+        '96.0.4664.55',
+        '96.0.4664.93',
+        '97.0.4692.20',
+    )
+    return _USER_AGENT_TPL % random.choice(_CHROME_VERSIONS)
+
+
+USER_AGENTS = {
+    'Safari': 'Mozilla/5.0 (X11; Linux x86_64; rv:10.0) AppleWebKit/533.20.25 (KHTML, like Gecko) Version/5.0.4 Safari/533.20.27',
+}
+
+
+# Use make_std_headers() to get a copy of these
+_std_headers = CaseInsensitiveDict({
+    'User-Agent': random_user_agent(),
+    'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8',
+    'Accept-Language': 'en-us,en;q=0.5',
+    'Sec-Fetch-Mode': 'navigate',
+})
+
+
+def handle_youtubedl_headers(headers):
+    filtered_headers = headers
+
+    if 'Youtubedl-no-compression' in filtered_headers:
+        filtered_headers = {k: v for k, v in filtered_headers.items() if k.lower() != 'accept-encoding'}
```

Comment:
sort of, as long as we make a copy of the headers.

This is only used by external downloaders now.

However I notice that the header, proxy etc. handling for external downloaders is... really bad. There is inconsistent handling of these special headers and proxies. But if we were to improve it there might be a lot of overlap with what RequestHandler does when preparing Request objects, so not sure how we should approach that.

Reply:
> However I notice that the header, proxy etc. handling for external downloaders is... really bad. There is inconsistent handling of these special headers and proxies.

pls elaborate

Reply:
Having a better look, it's not too bad as I initially thought. More that there is a lack of sanitization in some areas, but not that serious. It's out of scope of this networking rework project but something that could be improved upon sometime.

some things from a quick read:
- only FFmpegFD uses handle_youtubedl_headers to strip accept-encoding header (and the no compression internal header). Other downloaders leave both in and leak to server.
- Other internal headers will be leaked if used (though not much of a worry since they are rarely used)
-  passing proxies is only supported by a few external downloaders (and if they don't they will use no proxy without any sort of notice to user)
     - this is also an issue with current websocket support
- external downloaders that do support proxies generally just take it from ydl opts 
    - ignoring env proxies, though depends on if the external downloader will pick those up itself
- FFmpegFD proxy sanitization will create a bad proxy url if you use a scheme-less url like `//example.com:1080`
- Other external downloaders don't do any proxy validation/sanitization
- yt-dlp doesn't support `socks5h` scheme, rather it treats `socks5` scheme as `socks5h` (use remote DNS). Downloaders that do separate the two will treat the socks5 proxy as the one that uses local DNS (i.e. potential unwanted DNS leak)
- user passes proxy supported by yt-dlp but not by the external downloader